### PR TITLE
Moved lint to seperate job. Some refactor as well.

### LINKED ISSUE #1248: GitHub Actions: Find and implement fail-fast strategy
**Description of issue or feature request**:

#1242 added a GitHub Actions based CI workflow, which runs all configured builds in parallel, ignoring each others build results. However, GitHub Actions provides an option (i.e. `fail-fast`), to abort all builds if any build fails. This option might be useful for specific build relationships, where one failing build is likely to entail other failing builds.

**Current behavior**:
`fail-fast: false`

**Expected behavior**:
Consider `fail-fast: true` (default) but with specific conditions, e.g. prioritize `lint` build and abort any other build if that one fails.

### PR DESCRIPTION
**Pull request by shabeebk**
Please fill in the fields below to submit a pull request.  The more information
that is provided, the better.

Fixes #1248 

**Description of the changes being introduced by the pull request**:
The CI pipeline has been changed to run the lint tasks in the first and fail if the lint job fails.

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by shabeebk** [comment]
@lukpueh @jku, thanks for spending time reviewing the works :+1:. I have got a new PR with cleaner commits and signoff. Please review. The requirement is achieved and the suggestions are taken care of.

**Comment by lukpueh** [comment]
FYI: You wouldn't have needed to submit new PRs for a cleaner git history. You can change your git history locally, e.g. using `git rebase`, and then force push to your remote branch, e.g. `git push origin issue-1248-2 -f`.

Our [git history guideline doc](https://github.com/secure-systems-lab/lab-guidelines/blob/master/git-history.md) talks a bit about rebasing on an open PR, and has some pointers to good further reading.

File: .github/workflows/_test.yml [hunk 1]
```diff
@@ -5,24 +5,50 @@ on:
...
   tests:
     name: Tests
+    needs: lint-test
+    continue-on-error: ${{ matrix.failure-allowed }}
     strategy:
-      fail-fast: false
-      # Run regular TUF tests on each OS/Python combination, plus special tests
-      # (sslib main) and linters on Linux/Python3.x only.
+      # Run regular TUF tests on each OS/Python combination, plus
+      # sslib-master on Linux/Python3.x only.
```
Comment (lukpueh) [inline .github/workflows/_test.yml]: Looks like you undid this change we merged in yesterday via #2315, while resolving a conflict? Please change back:
```suggestion
      # (sslib main) on Linux/Python3.x only.
```

File: .github/workflows/_test.yml [hunk 2]
```diff
@@ -5,24 +5,50 @@ on:
 permissions: {}
 
 jobs:
+  lint-test:
+    name: Lint Test
+    runs-on: ubuntu-latest
+
+    env:
+      # Set TOXENV env var to tell tox which testenv (see tox.ini) to use
+      TOXENV: lint
```
Comment (lukpueh) [inline .github/workflows/_test.yml]: You can also configure toxenv via command line argument, then you don't need to set an envvar.
```suggestion
```

File: .github/workflows/_test.yml [hunk 3]
```diff
@@ -5,24 +5,50 @@ on:
...
+          python-version: 3.x
+          cache: 'pip'
+          cache-dependency-path: 'requirements/*.txt'
+
+      - name: Install dependencies
+        run: |
+          python3 -m pip install --constraint requirements/build.txt tox coveralls
+
+      - name: Run tox (${{ env.TOXENV }})
+        run: tox
```
Comment (lukpueh) [inline .github/workflows/_test.yml]: You can also configure toxenv via command line argument, then you don't need to set an envvar.
```suggestion
      - name: Run tox
        run: tox -e lint
```

File: .github/workflows/_test.yml [hunk 4]
```diff
@@ -5,24 +5,50 @@ on:
...
+
   tests:
     name: Tests
+    needs: lint-test
+    continue-on-error: ${{ matrix.failure-allowed }}
```
Comment (lukpueh) [inline .github/workflows/_test.yml]: No need to use a variable here, if it always has the same value
```suggestion
    continue-on-error: true
```

File: .github/workflows/_test.yml [hunk 5]
```diff
@@ -5,24 +5,50 @@ on:
...
-      # Run regular TUF tests on each OS/Python combination, plus special tests
-      # (sslib main) and linters on Linux/Python3.x only.
+      # Run regular TUF tests on each OS/Python combination, plus
+      # sslib-master on Linux/Python3.x only.
       matrix:
+        toxenv: [py]
         python-version: ["3.7", "3.8", "3.9", "3.10", "3.11"]
         os: [ubuntu-latest, macos-latest, windows-latest]
-        toxenv: [py]
+        failure-allowed: [true]
```
Comment (lukpueh) [inline .github/workflows/_test.yml]: No need to use a variable here, if it always has the same value
```suggestion
```

File: .github/workflows/_test.yml [hunk 6]
```diff
@@ -5,24 +5,50 @@ on:
...
         include:
           - python-version: 3.x
             os: ubuntu-latest
             toxenv: with-sslib-main
             experimental: true
-          - python-version: 3.x
-            os: ubuntu-latest
-            toxenv: lint
+            failure-allowed: true
```
Comment (lukpueh) [inline .github/workflows/_test.yml]: No need to use a variable here, if it always has the same value
```suggestion
```

**Review by lukpueh — CHANGES_REQUESTED** [review]
Thanks! I left a few more suggestions. Please accept, and we can merge.

**Comment by shabeebk** [comment]
@lukpueh, accepted. All look good to me. :+1:

**Review by lukpueh — APPROVED** [review]
LGTM, thanks!

**Pull request merged**